File: source/CSharp/Newton.cs
using System;

namespace Exploring.Net.CSharp {
    public static class Newton {
        public static double SquareRoot(this double input) {
            if (input < 0.0) throw new ArgumentOutOfRangeException();
            if (input == 0.0) return input;
            return FindFixedPoint(guess => (guess + input / guess) / 2.0, 1.0);
        }

        public static double SquareRoot2(double input) {
            return FindRoot(x => x * x - input, 1.0);
        }

        public static double FindRoot(Func<double, double> function, double guess) {
            Func<double, double> newtonsTransform = x => x - function(x) / Derivative(function)(x);
            return FindFixedPoint(newtonsTransform, guess);
        }

        static Func<double, double> Derivative(Func<double, double> function) {
            return x => (function(x + smallAmount) - function(x)) / smallAmount;
        }

        static double FindFixedPoint(Func<double, double> improveGuess, double initialGuess) {
            var guess = initialGuess;
            while (true) {
                var newGuess = improveGuess(guess);
                if (double.IsNaN(newGuess) || double.IsInfinity(newGuess)) return newGuess;
                if (CloseEnough(guess, newGuess)) return newGuess;
                guess = newGuess;
            }
        }

        static bool CloseEnough(double guess, double newGuess) {
            return Math.Abs((guess - newGuess) / guess) < smallAmount;
        }

        const double smallAmount = 0.000001;
    }
}

File: source/FSharp/Newton.fs
namespace Exploring.Net.FSharp

open System

module Newton =
    let smallAmount = 1E-6

    let rec findFixedPoint improveGuess guess =
        let newGuess = improveGuess guess
        if Double.IsNaN newGuess || Double.IsInfinity newGuess
            then newGuess
            else
                let goodEnough = abs((guess - newGuess) / guess) < smallAmount
                if goodEnough then newGuess else findFixedPoint improveGuess newGuess 

    let squareRoot input =
        let squareRootFromGuess = findFixedPoint (fun guess -> (guess + input / guess) / 2.0)
        match input with
            | 0.0 -> 0.0
            | input when input < 0.0 -> raise (new System.ArgumentOutOfRangeException())
            | input -> squareRootFromGuess 1.0 

    let findRoot aFunction firstGuess =
        let derivative = fun x -> (aFunction (x + smallAmount) - aFunction x) / smallAmount
        let newtonsTransform = fun x -> x - (aFunction x) / (derivative x)
        findFixedPoint newtonsTransform firstGuess

    let squareRoot2 input = findRoot (fun x -> x*x - input) 1.0
